pkg/util/stop/stopper.go

```diff
@@ -341,13 +341,24 @@ func (s *Stopper) RunAsyncTask(
 		return ErrUnavailable
 	}
 
-	ctx, span := tracing.ForkCtxSpan(ctx, taskName)
+	var sp *tracing.Span
+	if parentSp := tracing.SpanFromContext(ctx); parentSp != nil {
+		// NB: we don't use auto collection because it does not make
+		// sense for spans that extend past the lifetime of their
+		// parent. Nobody will collect the recording, but setting up
+		// a span is still helpful when an external tracer is set up.
+		ctx, sp = parentSp.Tracer().StartSpanCtx(ctx, taskName, tracing.WithParentAndManualCollection(parentSp.Meta()))
```

Comment:
@irfansharif this is actually obsolete with the other commit, which teaches `ForkCtxSpan` to do the right thing.

Unfortunately a lot of tracing-based tests break because they use this pattern:

```go
sp := tr.StartSpan("foo")
stopper.RunAsyncTask("bar", func(ctx context.Context) {
  log.Event(ctx, "bar")
}
if !contains(sp.Recording(), "bar") { fail }
```

This is no longer reasonable because the span that contains `bar` is not a direct child of `sp`. Instead, I think such tests really want to use tracing as it is supposed to be used, where all spans are recorded somewhere and can be retrieved.

```go
type RecordingCollector struct {
  mu struct {
    spans []*Span
  }
}

var rc tracing.RecordingCollector
sp := tr.StartSpan("foo", tracing.WithRecordingCollector(&rc))
// Child spans will all (also?) report to that collector, whether direct or through .Meta()

// ...
require.Contains(t, rc.GetRecording(), "foo")
```

Of course there are limits to in-memory collectors. They can't travel across process/RPC boundaries. So while `Meta()` could hold a reference to `rc` above, if we serialized it through an RPC we could not propagate it to the other end. But you could imagine collectors that could do it, for example a collector that is just an endpoint to which to send the recordings. That seems fine; we just need the collector to be able to tell the tracer how to serialize (if at all). There could even be hybrids: a local in-mem recording collector would serialize as an "send to this endpoint" collector.

The collector business would also replace the `WithParentAnd{Auto,Manual}Collection` options. The auto mode would be a `WithParentCollector` (i.e. report to the parent directly) and the manual mode is basically - no collection, and would be the default. The parent and the collection mode would be split, i.e. we'd add `WithParent` and `WithParentMeta`.

Curious what you think and whether there is a less invasive option we could use. We do need these tests to pass, unfortunately, but the changes in this PR are imo very worthwhile. We still leak 25mb of memory on long-running nodes because some long-running spans hold 1000 spans alive.

Reply:
If we take it a step further, there's even a way to make tracing more idiomatic and do away with that manual `ImportRemoteSpans` stuff altogether.

Conceptually, what I would always have preferred is if the gateway node (or wherever the root span is) opened some kind of mailbox, and then all of the spans would report there. The problem I always saw with that was that you couldn't really tell when to stop waiting for new spans. But I think the problems we're seeing with the spans-as-record approach are pretty annoying too (think about how easy memory management could be if spans didn't hold on to each other and messages, etc, but just shipped things out to a central location!). Really the only benefit they give us is that the trace is done the moment the requests have executed and the root span finishes.

I don't think this can happen now, but it's worth keeping in mind as an ultimate goal. Curious what @andreimatei thinks.